docs/0001-teleport-job-worker.md

```diff
@@ -0,0 +1,185 @@
+  ```
+* **Query**
+  ```shell
+  job-worker -token "some-token" query -id "some-job-id"
+  ```
+  Returns details about the job that matches the provided ID.
+* **Output**
+  ```shell
+  job-worker -token "some-token" output -id "some-job-id"
+  ```
+  Returns all the output for the job that matches the provided ID.
+
+### Process Execution
+
+Jobs will be forked from the service's process and contain exactly one process that it manages. A job must have one of
+the statuses defined in `Statuses` of the [proto spec](../api/proto/job/job.proto) and will transition between statuses
+as show below:
+
+```
+  ┌───────────────────────────┐
+  │                           │
+ready      ┌────►stopped◄─────┤
+  │        │                  │
+  │        │                  │
+  │        ├────►failed◄──────┘
+  │        │                   
+  ▼        │                   
+running────┴────►succeeded     
+```
+
+Resource control will be implemented using cgroups. A parent directory `job-worker` will be created
+under `/sys/fs/cgroup` that will contain a directory for every job created. For example, the path for `job-a` would look
+like `/sys/fs/cgroup/job-worker/job-a`. Under this path, the following files would be updated to manage the resources
+for the job:
+
+* `cgroup.procs` contains the PID of the job's process being managed
+* `cpu.max` limits CPU usage
+* `memory.max` limits memory usage
+* `io.max` limits disk IO
+
+Note that for this implementation, job metadata will remain in memory until the application exits. A real implementation
+of this service would store job metadata in some kind of persistent storage like a database.
+
+### Proto Specification
+
+The service will implement the following protobufs:
+
+* [job](../api/proto/job/job.proto)
+
+### API
+
+The service's API will contain the resources listed below and be made available using gRPC.
+
+#### job
+
+A job is responsible for managing the lifecycle of the process given to it.
+
+* `Start` start a new job and begin execution of the specified command immediately
+* `Stop` stop execution of the specified job immediately
+* `Query` query details about specified job; this function can run on a job of any status
+* `Output` get the full output (stdout and stderr) of any existing job; the output from the job will be written to
+  memory as the process executes so that it can be retrieved by calls to this function; for this exercise, output will
+  remain available until the job worker has exited; for real world uses, the output would need to be removed or written
+  to persistent storage because of memory limitations
+
+Details of a job's functions are listed in [job.proto](../api/proto/job/job.proto).
+
+### Library Godoc
+```
+package library // import "github.com/kurczynski/teleport-job-worker/pkg/library"
+
+TYPES
+
+type Job struct {
+	// Has unexported fields.
+}
+    Job Contains information to interact with jobs.
+
+func GetJob(id string) *Job
```

Comment:
That's what I had in mind since we want to query the jobs. I guess an important question would be, what kind of information do we want returned from a query call?

Reply:
I would avoid global state in the library.

While the API needs to keep a store of jobs, I don't see a similar requirement for the library.